gnosis-imap/gnosis-imap-spec.ini

```diff
@@ -68,7 +68,119 @@ localMem: M
+    CEIL_DATA_LEN +Int 512, 32, _),
+    CEIL_DATA_LEN +Int 160, 32, _),
+    64, 32, _),
+  
+    ;start of result, divided into 4 parts:
+    ;return abi.encodePacked(byte(0x19), byte(0x01), domainSeparator, safeTxHash)
+    CEIL_DATA_LEN +Int 576, 32, #padRightToWidth(32, #parseByteStack("0x19")) ),
+    CEIL_DATA_LEN +Int 577, 32, #padRightToWidth(32, #parseByteStack("0x01")) ),
+    CEIL_DATA_LEN +Int 578, 32, #buf(32, DOMAIN_SEPARATOR) ),
+    CEIL_DATA_LEN +Int 610, 32, #buf(32, SAFE_TX_HASH ) ),
+    ;output len
+    CEIL_DATA_LEN +Int 544, 32, #padToWidth(32, #asByteStack(66)) ),
+    ;next free memory loc
+    64, 32, #buf(32, CEIL_DATA_LEN +Int 642) )
+pc: 16193 => 16786
+gas: #gas(STARTGAS,
+          X                                                => (X +Int 6 *Int ((DATA_LEN +Int 31) /Int 32)) +Int 941,
+          Cmem(BYZANTIUM, START_MEM_USED => END_MEM_USED ) )
+memoryUsed: START_MEM_USED => END_MEM_USED
+log: _ => _
+refund: _ => _
+coinbase: _ => _
+proxy_storage:
+    store(store(M, #hashedLocation({COMPILER}, {MASTER_COPY}, .IntList), MASTER_COPY_ID),
+                   #hashedLocation({COMPILER}, {DOMAIN_SEPARATOR}, .IntList), DOMAIN_SEPARATOR)
++requires:
+    ; Range
+    andBool #rangeAddress(MSG_SENDER)
+
+    andBool #rangeAddress(          TO)
+    andBool #rangeUInt( 256,        VALUE)
+    andBool #rangeUInt( 256,        DATA_LEN)
+    ;andBool #rangeBytes( DATA_LEN,  DATA)
+    andBool #rangeUInt(   8,        OPERATION)
+    ; enum Enum.Operation, 3 possible values encoded to 0-2.
+    andBool OPERATION <=Int 2
+    andBool #rangeUInt( 256,        SAFE_TX_GAS)
+    andBool #rangeUInt( 256,        DATA_GAS)
+    andBool #rangeUInt( 256,        GAS_PRICE)
+    andBool #rangeAddress(          GAS_TOKEN)
+    andBool #rangeAddress(          REFUND_RECEIVER)
+    andBool #rangeUInt( 256,        NONCE)
+
+    andBool #rangeAddress(          MASTER_COPY_ID)
+    andBool #rangeBytes( 32,        DOMAIN_SEPARATOR)
+
+    ; TODO more precise value. 1000 is too little.
+    andBool #range(0 <= #sizeWordStack(WS) <= 900)
+    andBool #range(0 <= MU < pow256)
+
+    andBool DATA_LEN <Int 2 ^Int 16
+
+    andBool SAFE_TX_HASH ==Int keccak( #encodeArgs(
+                                 #bytes32(#parseHexWord({SAFE_TX_TYPEHASH})),
+                                 #address(TO),
+                                 #uint256(VALUE),
+                                 #bytes32(keccak(#buf(DATA_LEN, DATA))),
+                                 #uint8(OPERATION),
+                                 #uint256(SAFE_TX_GAS),
+                                 #uint256(DATA_GAS),
+                                 #uint256(GAS_PRICE),
+                                 #address(GAS_TOKEN),
+                                 #address(REFUND_RECEIVER),
+                                 #uint256(NONCE) ))
+    andBool CEIL_DATA_LEN ==Int #ceil32(DATA_LEN)
+    andBool INIT_MEM ==K   storeRange( storeRange( storeRange( storeRange( M, 
```

Comment:
do you mean the `M` in the proxy_storage?

Reply:
oops, good catch:) Consequence of copy-paste.